File: iOS/Utils/Utils/Utils/FileUtils.m
#import "FileUtils.h"

@implementation FileUtils

+(NSString *) getDocumentPath {
    NSArray * paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSAllDomainsMask, YES);
    return paths[0];
}

+(BOOL) makeDir: (NSString *)dir {
    BOOL ret = NO;
    NSFileManager * fmgr = [NSFileManager defaultManager];
    if (![fmgr fileExistsAtPath: dir]) {
        ret = [fmgr createDirectoryAtPath:dir withIntermediateDirectories:YES attributes:nil error:nil];
    }
    return ret;
}

+(NSString *) makeDir: (NSString *)dir basePath: (NSString *)path {
    BOOL ret = NO;
    NSString * pathTmp = [NSString stringWithFormat:@"%@/%@", path, dir];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    if (![fmgr fileExistsAtPath: pathTmp]) {
        ret = [fmgr createDirectoryAtPath:pathTmp withIntermediateDirectories:YES attributes:nil error:nil];
    }
    return pathTmp;
}

+(NSString *) buildPath: (NSString *)fileName path: (NSString *)path {
    NSString * document = [self getDocumentPath];
    NSString * pathTmp = [self makeDir: path basePath:document];
    NSString * fileOper = [NSString stringWithFormat:@"%@/%@", pathTmp, fileName];
    return fileOper;
}

+(NSString *) getLocalFile: (NSString *)name type: (NSString *)type {
    return [[NSBundle mainBundle] pathForResource:name ofType:type];
}

+(NSString *) getLocalFile: (NSString *)name type: (NSString *)type dir: (NSString *)dir {
    return [[NSBundle mainBundle] pathForResource:name ofType:type inDirectory:dir];
}

+(void) rewriteTextFile:(NSString *)fileName savePath: (NSString *)path fileContent: (NSString *)text {
    NSString * fileOper = [self buildPath:fileName path: path];
    NSData * writeData = [text dataUsingEncoding: NSUTF8StringEncoding];
    [writeData writeToFile:fileOper atomically:YES];
}

+(void) appendTextFile: (NSString *)fileName savePath: (NSString *)path fileContent: (NSString *)text {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSFileHandle * file = [NSFileHandle fileHandleForWritingAtPath:fileOper];
    if (file != nil) {
        [file seekToEndOfFile];
        NSData * writeData = [text dataUsingEncoding: NSUTF8StringEncoding];
        [file writeData: writeData];
        [file closeFile];
    } else {
        [self rewriteTextFile:fileName savePath:path fileContent:text];
    }
}

+(NSString *) readTextFile: (NSString *)fileName loadPath: (NSString *)path {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSMutableString * text = [[NSMutableString alloc] initWithString:@""];
    if ([self fileExists:fileName filePath:path]) {
        NSData * readData = [NSData dataWithContentsOfFile:fileOper];
        text = [[NSMutableString alloc] initWithData:readData encoding:NSUTF8StringEncoding];
    }
    return text;
}

+(void) rewriteFile:(NSString *)fileName savePath: (NSString *)path fileData: (NSData *)data {
    NSString * fileOper = [self buildPath:fileName path:path];
    [data writeToFile:fileOper atomically:YES];
}

+(void) appendFile: (NSString *)fileName savePath: (NSString *)path fileData: (NSData *)data {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSFileHandle * file = [NSFileHandle fileHandleForWritingAtPath:fileOper];
    if (file != nil) {
        [file seekToEndOfFile];
        [file writeData: data];
        [file closeFile];
    } else {
        [self rewriteFile:fileName savePath:path fileData:data];
    }
}

+(NSData *) readFile: (NSString *)fileName loadPath: (NSString *)path {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSData * retData = nil;
    if ([self fileExists:fileName filePath:path]) {
        retData = [NSData dataWithContentsOfFile:fileOper];
    }
    return retData;
}

+(BOOL) deleteFile: (NSString *)fileName path: (NSString *)path {
    BOOL ret = NO;
    NSString * fileOper = [self buildPath:fileName path:path];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    ret = [fmgr removeItemAtPath:fileOper error:nil];
    return ret;
}

+(BOOL) deleteDir: (NSString *)dir path: (NSString *)path {
    BOOL ret = NO;
    NSString * fileOper = [self buildPath:dir path:path];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    ret = [fmgr removeItemAtPath:fileOper error:nil];
    return ret;
}

+(BOOL) copyFile: (NSString *)sourceFile sourcePath: (NSString *)sourcePath destFile: (NSString *)destFile destPath: (NSString *)destPath {
    BOOL ret = NO;
    NSString * fileSource = [self buildPath:sourceFile path:sourcePath];
    NSString * fileDest = [self buildPath:destFile path:destPath];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    ret = [fmgr copyItemAtPath:fileSource toPath:fileDest error:nil];
    return ret;
}

+(BOOL) copyFolder: (NSString *)sourceFolder sourcePath: (NSString *)sourcePath destFolder: (NSString *)destFolder destPath: (NSString *)destPath {
    BOOL ret = NO;
    NSString * fileSource = [self buildPath:sourceFolder path:sourcePath];
    NSString * fileDest = [self buildPath:destFolder path:destPath];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    ret = [fmgr copyItemAtPath:fileSource toPath:fileDest error:nil];
    return ret;
}

+(BOOL) moveFile: (NSString *)sourceFile sourcePath: (NSString *)sourcePath destFile: (NSString *)destFile destPath: (NSString *)destPath {
    BOOL ret = NO;
    NSFileManager * fmgr = [NSFileManager defaultManager];
    NSString * fileSource = [self buildPath:sourceFile path:sourcePath];
    NSString * fileDest = [self buildPath:destFile path:destPath];
    ret = [fmgr moveItemAtPath:fileSource toPath:fileDest error:nil];
    return ret;
}

+(BOOL) moveFolder: (NSString *)sourceFolder sourcePath: (NSString *)sourcePath destFolder: (NSString *)destFolder destPath: (NSString *)destPath {
    BOOL ret = NO;
    NSFileManager * fmgr = [NSFileManager defaultManager];
    NSString * fileSource = [self buildPath:sourceFolder path:sourcePath];
    NSString * fileDest = [self buildPath:destFolder path:destPath];
    ret  =[fmgr moveItemAtPath:fileSource toPath:fileDest error:nil];
    return ret;
}

+(NSArray *) loadArrayFromFile: (NSString *)fileName path: (NSString *)path {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSArray * arr = [NSArray arrayWithContentsOfFile:fileOper];
    return arr;
}

+(BOOL) saveArrayToFile: (NSString *)fileName path: (NSString *)path array: (NSArray *)array {
    BOOL ret = NO;
    if (array != nil) {
        NSString * fileOper = [self buildPath:fileName path:path];
        ret = [array writeToFile:fileOper atomically:YES];
    }
    return ret;
}

+(BOOL) fileExists: (NSString *)fileName filePath: (NSString *)path {
    NSString * fileOper = [self buildPath:fileName path:path];
    NSFileManager * fmgr = [NSFileManager defaultManager];
    return [fmgr fileExistsAtPath: fileOper];
}

+(BOOL) fileExists: (NSString *)filePath {
    return [[NSFileManager defaultManager] fileExistsAtPath: filePath];
}

+(NSInteger) getFileSize: (NSString *)filePath {
    NSFileManager * fmgr = [NSFileManager defaultManager];
    if ([fmgr fileExistsAtPath: filePath]) {
        NSDictionary * attrs = [fmgr attributesOfItemAtPath:filePath error:nil];
        return (NSInteger)[attrs fileSize];
    }
    return 0;
}

+(NSInteger) getDirSize: (NSString *)folderPath {
    NSFileManager * fmgr = [NSFileManager defaultManager];
    if (![fmgr fileExistsAtPath: folderPath]) {
        return 0;
    }
    NSArray * childFilesEnumerator = [fmgr subpathsAtPath: folderPath];
    NSInteger folderSize = 0;
    if (childFilesEnumerator != nil) {
        for (NSString * fileName in childFilesEnumerator) {
            NSString * fileAbsolutePath = [NSString stringWithFormat:@"%@/%@", folderPath, fileName];
            folderSize += [self getFileSize: fileAbsolutePath];
        }
    }
    return folderSize;
}

+(NSString *) getReadableFileSize: (NSInteger)fileSize {
    double fz = fileSize / 1024.0;
    return [NSString stringWithFormat:@"%f KB", fz];
}


@end

File: Swift/Utils/FileUtils.swift
import UIKit

class FileUtils: NSObject {
    
    class func getDocumentPath() -> String {
        var paths = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.AllDomainsMask, true);
        return paths[0] as String
    }
    
    class func makeDir(dir: String) -> Bool {
        var ret = false
        var fmgr = NSFileManager.defaultManager()
        if (!fmgr.fileExistsAtPath(dir)) {
            ret = fmgr.createDirectoryAtPath(dir, withIntermediateDirectories: true, attributes: nil, error: nil)
        }
        return ret
    }
    
    class func makeDir(dir: String, basePath path: String) -> String {
        var ret = false
        var pathTmp = "\(path)/\(dir)"
        var fmgr = NSFileManager.defaultManager()
        
        if (!fmgr.fileExistsAtPath(pathTmp)) {
            ret = fmgr.createDirectoryAtPath(pathTmp, withIntermediateDirectories:true, attributes:nil, error:nil)
        }
        return pathTmp
    }
    
    class func buildPath(fileName: String, path: String) -> String {
        var document = self.getDocumentPath()
        var pathTmp = self.makeDir(path, basePath:document)
        var fileOper = "\(pathTmp)/\(fileName)"
        return fileOper
    }
    
    class func getLocalFile(name: String, type: String) -> String? {
        return NSBundle.mainBundle().pathForResource(name, ofType: type)
    }
    
    class func getLocalFile(name: String, type: String, dir: String) -> String? {
        return NSBundle.mainBundle().pathForResource(name, ofType: type, inDirectory: dir)
    }
    
    class func rewriteTextFile(fileName: String, savePath path: String, fileContent text: String) {
        var fileOper = buildPath(fileName, path: path)
        var writeData = (text as NSString).dataUsingEncoding(NSUTF8StringEncoding)
        writeData!.writeToFile(fileOper, atomically:true)
    }
    
    class func appendTextFile(fileName: String, savePath path: String, fileContent text: String) {
        var fileOper = buildPath(fileName, path: path)
        var file = NSFileHandle(forWritingAtPath: fileOper)
        if (file != nil) {
            file!.seekToEndOfFile()
            var writeData = (text as NSString).dataUsingEncoding(NSUTF8StringEncoding)
            file!.writeData(writeData!)
            file!.closeFile()
        } else {
            rewriteTextFile(fileName, savePath: path, fileContent: text)
        }
    }
    
    class func readTextFile(fileName: String, loadPath path: String) -> String {
        var fileOper = buildPath(fileName, path: path)
        var text: String = ""
        if (self.fileExists(fileName, filePath:path)) {
            var readData = NSData(contentsOfFile: fileOper)
            text = NSString(data: readData!, encoding: NSUTF8StringEncoding) as String
        }
        return text
    }
    
    class func rewriteFile(fileName: String, savePath path: String, fileData data: NSData) {
        var fileOper = buildPath(fileName, path: path)
        data.writeToFile(fileOper, atomically:true)
    }
    
    class func appendFile(fileName: String, savePath path: String, fileData data: NSData) {
        var fileOper = buildPath(fileName, path: path)
        var file = NSFileHandle(forWritingAtPath: fileOper)
        if (file != nil) {
            file!.seekToEndOfFile()
            file!.writeData(data)
            file!.closeFile()
        } else {
            rewriteFile(fileName, savePath: path, fileData: data)
        }
    }
    
    class func readFile(fileName: String, loadPath path: String) -> NSData? {
        var fileOper = buildPath(fileName, path: path)
        var retData: NSData? = nil
        if (self.fileExists(fileName, filePath:path)) {
            retData = NSData(contentsOfFile: fileOper)
        }
        return retData
    }
    
    class func deleteFile(fileName: String, path: String) -> Bool {
        var ret = false
        var fileOper = buildPath(fileName, path: path)
        var fmgr = NSFileManager.defaultManager()
        ret = fmgr.removeItemAtPath(fileOper, error: nil)
        return ret
    }
    
    class func deleteDir(dir: String, path: String) -> Bool {
        var ret = false
        var fileOper = buildPath(dir, path: path)
        var fmgr = NSFileManager.defaultManager()
        ret = fmgr.removeItemAtPath(fileOper, error: nil)
        return ret
    }
    
    class func copyFile(sourceFile: String, sourcePath: String, destFile: String, destPath: String) -> Bool {
        var ret = false
        var fileSource = buildPath(sourceFile, path: sourcePath)
        var fileDest = buildPath(destFile, path: destPath)
        var fmgr = NSFileManager.defaultManager()
        ret = fmgr.copyItemAtPath(fileSource, toPath: fileDest, error: nil)
        return ret
    }
    
    class func copyFolder(sourceFolder: String, sourcePath: String, destFolder: String, destPath: String) -> Bool {
        var ret = false
        var fileSource = buildPath(sourceFolder, path: sourcePath)
        var fileDest = buildPath(destFolder, path: destPath)
        var fmgr = NSFileManager.defaultManager()
        ret = fmgr.copyItemAtPath(fileSource, toPath: fileDest, error: nil)
        return ret
    }
    
    class func moveFile(sourceFile: String, sourcePath: String, destFile: String, destPath: String) -> Bool {
        var ret = false
        var fmgr = NSFileManager.defaultManager()
        var fileSource = buildPath(sourceFile, path: sourcePath)
        var fileDest = buildPath(destFile, path: destPath)
        ret = fmgr.moveItemAtPath(fileSource, toPath: fileDest, error: nil)
        return ret
    }
    
    class func moveFolder(sourceFolder: String, sourcePath: String, destFolder: String, destPath: String) -> Bool {
        var ret = false
        var fmgr = NSFileManager.defaultManager()
        var fileSource = buildPath(sourceFolder, path: sourcePath)
        var fileDest = buildPath(destFolder, path: destPath)
        ret = fmgr.moveItemAtPath(fileSource, toPath: fileDest, error: nil)
        return ret
    }
    
    class func loadArrayFromFile(fileName: String, path: String) -> NSArray? {
        var fileOper = buildPath(fileName, path: path)
        var arr = NSArray(contentsOfFile: fileOper)
        return arr
    }
    
    class func saveArrayToFile(fileName: String, path: String, array: NSArray?) -> Bool {
        var ret = false
        if (array != nil) {
            var fileOper = buildPath(fileName, path: path)
            ret = array!.writeToFile(fileOper, atomically: true)
        }
        return ret
    }
    
    class func fileExists(fileName: String, filePath path: String) -> Bool {
        var fileOper = buildPath(fileName, path: path)
        var fmgr = NSFileManager.defaultManager()
        return fmgr.fileExistsAtPath(fileOper)
    }
    
    class func fileExists(filePath: String) -> Bool {
        return NSFileManager.defaultManager().fileExistsAtPath(filePath)
    }
    
    class func getFileSize(filePath: String) -> UInt64 {
        var manager = NSFileManager.defaultManager()
        if (manager.fileExistsAtPath(filePath)) {
            var attrs = manager.attributesOfItemAtPath(filePath, error: nil)
            var dic = NSDictionary(dictionary: attrs!)
            return dic.fileSize()
        }
        return 0
    }
    
    class func getDirSize(folderPath: String) -> UInt64 {
        var manager = NSFileManager.defaultManager()
        if (!manager.fileExistsAtPath(folderPath)) {
            return 0
        }
        var childFilesEnumerator = manager.subpathsAtPath(folderPath)
        var folderSize: UInt64 = 0
        if (childFilesEnumerator != nil) {
            for fileName in childFilesEnumerator! {
                var fileAbsolutePath = "\(folderPath)/\(fileName)"
                folderSize += self.getFileSize(fileAbsolutePath)
            }
        }
        // Byte to M
        return folderSize
        
    }
    
    class func getReadableFileSize(fileSize: UInt64) -> String {
        var fz: Double = Double(fileSize)
        return "\(fz / 1024.0) KB"
    }
   
}
